### PR TITLE
Feature#31 employee check in out

### LINKED ISSUE #31: Employee Check In/Out
# REQUIREMENTS
- Button Check In should add a new record to the table attendance
- Button Check Out should add the time to the record that match the date and is null

### PR DESCRIPTION
**Pull request by AngelBFdev**
# Description
- Check In/Out button saves in the table attendance
- Validates Employee exists in the table Users
- Validates that Employee choose the correct company

### CODE REVIEW

File: app/controllers/attendances_controller.rb [hunk 1]
```diff
@@ -1,4 +1,40 @@
 class AttendancesController < ApplicationController
   def index
   end
+
+  def new
+    @attendance = Attendance.new
+    @company = Company.where(disable: false)
```
Comment (kevinnio) [inline app/controllers/attendances_controller.rb]: What about adding a scope to `Company` instead?

```ruby
# company.rb
class Company < ApplicationRecord
  # ...

  scope :enabled, -> { where(disabled: false) }
end 
```

After that, you can use it like this:

```ruby
@company = Company.enabled
```

https://guides.rubyonrails.org/active_record_querying.html#scopes

File: app/controllers/attendances_controller.rb [hunk 2]
```diff
@@ -1,4 +1,40 @@
...
+  end
+
+  def create
+    @employee = Employee.where(disable:false)
```
Comment (kevinnio) [inline app/controllers/attendances_controller.rb]: Same here, what about adding a `enabled` scope to `Employee`?

File: app/controllers/attendances_controller.rb [hunk 3]
```diff
@@ -1,4 +1,40 @@
...
+    data = params.require(:attendance).permit(:private_num,:company_id)
+    time = Time.now.strftime("%H:%M:%S")
+    day = Time.now.strftime("%d-%m-%Y")
+    emp = @employee.find_by(private_num: data.fetch(:private_num), company_id: data.fetch(:company_id))
+
+    if !emp.nil?
```
Comment (kevinnio) [inline app/controllers/attendances_controller.rb:16]: This whole `if` block could be replaced with a guard clause instead:

```ruby
return if emp.nil?

# stuff to do if emp is not nil goes here.
```

https://anthonygharvey.com/guard-clauses-vs-nested-conditionals

File: app/controllers/pages_controller.rb [hunk 4]
```diff
@@ -1,4 +1,14 @@
 class PagesController < ApplicationController
   def home
+    @attendance = Attendance.new
+  end
+  def create
```
Comment (kevinnio) [inline app/controllers/pages_controller.rb]: Remember to leave a blank line between method definitions so your code is easy to read.

File: app/controllers/pages_controller.rb [hunk 5]
```diff
@@ -1,4 +1,14 @@
...
+    time = Time.now.strftime("%H:%M")
+    day = Time.now.strftime("%d-%m-%Y")
+
+    @attendance = Attendance.create(attendance_date: day, check_in: time, check_out: time, private_num: :private_num)
+    if @attendance.save
+      redirect_to pages_path
+    end
```
Comment (kevinnio) [inline app/controllers/pages_controller.rb]: There's a better way to write single-line `if`s in ruby:

```suggestion
    redirect_to pages_path if @attendance.save
```

File: db/migrate/20220129225619_remove_column.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+class RemoveColumn < ActiveRecord::Migration[6.1]
```
Comment (kevinnio) [inline db/migrate/20220129225619_remove_column.rb]: `RemoveColumn` doesn't tell me much about what this migration is trying to do. Remember to be descriptive with migrations names, for example:

- `CreateUsersTable`
- `AddUserRoleToUsersTable`
- `RemoveUserRoleFromUsersTable`
- etc

File: app/controllers/attendances_controller.rb [hunk 7]
```diff
@@ -1,4 +1,40 @@
 class AttendancesController < ApplicationController
   def index
   end
+
+  def new
+    @attendance = Attendance.new
+    @company = Company.where(disable: false)
+  end
+
+  def create
```
Comment (kevinnio) [inline app/controllers/attendances_controller.rb:10]: This method has too many lines. Usually, a method with more than 10 lines of code is considered too long in Ruby. You could try breaking it into smaller methods.

File: app/controllers/attendances_controller.rb [hunk 8]
```diff
@@ -1,4 +1,40 @@
...
+    emp = @employee.find_by(private_num: data.fetch(:private_num), company_id: data.fetch(:company_id))
+
+    if !emp.nil?
+      @all = Attendance.all
+      update = @all.find_by(private_num: data.fetch(:private_num), attendance_date: day, check_out: nil)
+      if update.nil?
+        data[:attendance_date]=day
+        data[:check_in]=time
+        data.delete(:company_id)
+        if Attendance.new(data).save
```
Comment (kevinnio) [inline app/controllers/attendances_controller.rb]:
```suggestion
        if Attendance.create(data)
```

**Review by kevinnio — CHANGES_REQUESTED** [review]
Left some comments

File: app/controllers/attendances_controller.rb (hunk 8 above)
Comment (AngelBFdev) [reply app/controllers/attendances_controller.rb]: Done

File: db/migrate/20220129225619_remove_column.rb (hunk 6 above)
Comment (AngelBFdev) [reply db/migrate/20220129225619_remove_column.rb]: Done

File: app/controllers/attendances_controller.rb (hunk 7 above)
Comment (AngelBFdev) [reply app/controllers/attendances_controller.rb:10]: Done

File: app/controllers/attendances_controller.rb [hunk 9]
```diff
@@ -1,4 +1,44 @@
...
+    @attendance = Attendance.new
+    @company = Company.enabled
+  end
+
+  def create
+    @employee = Employee.enabled
+    data = params.require(:attendance).permit(:private_num, :company_id)
+    data[:check_in] = Time.now.strftime("%H:%M:%S")
+    data[:attendance_date] = Time.now.strftime("%d-%m-%Y")
+    emp = @employee.find_by(private_num: data.fetch(:private_num), company_id: data.fetch(:company_id))
```
Comment (kevinnio) [inline app/controllers/attendances_controller.rb]: Now that I think about it, is there any reason to use `@employee` inside this action? Why not just do something like this instead:

```suggestion
    emp = Employee.enabled.find_by(data.slice(:private_num, :company_id))
```

**Review by kevinnio — CHANGES_REQUESTED** [review]
Nice! Almost there!

**Review by kevinnio — APPROVED** [review]
LGTM 👍

**Pull request merged**